File: sync/src/core/harness.ts
import { existsSync, readFileSync, statSync } from "node:fs";
import os from "node:os";
import path from "node:path";
import { assertNever, isErrno } from "@runtime/errors.ts";
import { ConfigProvider, Effect, HashMap, Schema } from "effect";
import {
  HARNESS_ADAPTERS,
  type HarnessAdapter,
  type HarnessHookSpec,
  type HarnessId,
  type HarnessLauncherSpec,
  type HostPlatform,
} from "./harness-adapters.ts";

export type { HarnessId, HostPlatform } from "./harness-adapters.ts";

export const SOURCE_AGENT_FILE = "HARNESS.md";
const DEFAULT_INSTRUCTION_FILE = "AGENTS.md";
const INSTALL_TIMEOUT_SECONDS = 120;
export const MANAGED_STATE_SUBDIR = ".local/share/agentium/sync-managed";
const DEFAULT_PACKAGE_CACHE_SUBDIR = ".local/share/agentium/pi-packages";
export const SKILLS_DST_DIR = "skills";
export const SKILLS_SOURCE_SUBDIR = "current";
const PATH_COMPONENT_PATTERN = /^[A-Za-z0-9._-]+$/;

export type HarnessHook =
  | {
      readonly kind: "PackageBootstrap";
      readonly manifestFile: string;
      readonly settingsFile: string;
      readonly cacheSubdir: string;
    }
  | {
      readonly kind: "ExtensionDeps";
      readonly rootDir: string;
    };

export interface HarnessSpec extends Omit<HarnessAdapter, "homeSegments" | "id" | "platforms"> {
  readonly id: HarnessId;
  readonly sourceName: string;
  readonly home: string;
}

export interface Harness {
  readonly id: HarnessId;
  readonly sourceName: string;
  readonly home: string;
  readonly launcher: Required<Omit<HarnessLauncherSpec, "env">> & {
    readonly env?: Record<string, string>;
  };
  readonly instructionFile: string;
  readonly runtimeSubdir: string | undefined;
  readonly compatManagedEntries: readonly string[];
  readonly hooks: readonly HarnessHook[];
}

export class SyncEnv {
  readonly home: string;
  readonly ssotHome: string;
  readonly dataHome: string;
  readonly skillsHome: string;
  readonly harnessesHome: string;
  readonly mcporterHome: string;
  readonly summarizeHome: string;
  readonly managedStateHome: string;
  readonly installTimeoutMs: number;
  readonly harnesses: readonly Harness[];
  readonly platform: HostPlatform;
  readonly rootEnv: Readonly<Record<string, string>>;

  constructor(
    home: string,
    ssotHome: string,
    dataHome: string,
    skillsHome: string,
    harnessesHome: string,
    mcporterHome: string,
    summarizeHome: string,
    managedStateHome: string,
    installTimeoutMs: number,
    harnesses: readonly Harness[],
    platform: HostPlatform = platformFromProcess(),
  ) {
    this.home = home;
    this.ssotHome = ssotHome;
    this.rootEnv = Effect.runSync(loadRootEnv(path.join(ssotHome, ".env")));
    this.dataHome = dataHome;
    this.skillsHome = skillsHome;
    this.harnessesHome = harnessesHome;
    this.mcporterHome = mcporterHome;
    this.summarizeHome = summarizeHome;
    this.managedStateHome = managedStateHome;
    this.installTimeoutMs = installTimeoutMs;
    this.harnesses = harnesses;
    this.platform = platform;
  }

  static fromSystem(): SyncEnv {
    const homeCandidates = [Bun.env["HOME"], process.env["HOME"], os.homedir()];
    const home = homeCandidates.find(
      (candidate): candidate is string =>
        typeof candidate === "string" && candidate.trim().length > 0,
    );
    if (!home) {
      throw new Error("missing HOME");
    }
    return SyncEnv.fromHome(home, INSTALL_TIMEOUT_SECONDS * 1000);
  }

  static fromHome(
    home: string,
    installTimeoutMs: number,
    options: {
      readonly platform?: HostPlatform;
    } = {},
  ): SyncEnv {
    const agentsHome = path.join(home, ".config", "agents");
    const dataHome = path.join(home, ".local", "share", "agentium");
    const harnessesHome = path.join(agentsHome, "harnesses");
    const platform = options.platform ?? platformFromProcess();
    return new SyncEnv(
      home,
      agentsHome,
      dataHome,
      path.join(agentsHome, "skills"),
      harnessesHome,
      path.join(home, ".mcporter"),
      path.join(home, ".summarize"),
      path.join(home, MANAGED_STATE_SUBDIR),
      installTimeoutMs,
      discoverHarnesses(home, harnessesHome, platform),
      platform,
    );
  }

  harness(id: HarnessId): Harness | undefined {
    return this.harnesses.find((harness) => harness.id === id);
  }
}

export class RootEnvReadError extends Schema.TaggedError<RootEnvReadError>()("RootEnvReadError", {
  path: Schema.String,
  cause: Schema.Unknown,
}) {
  override get message(): string {
    return `failed to read root environment file ${this.path}`;
  }
}

const readRootEnvContent = (envPath: string): Effect.Effect<string | undefined, RootEnvReadError> =>
  Effect.try({
    try: () => readFileSync(envPath, "utf-8"),
    catch: (cause) => ({ cause, isEnoent: isErrno(cause, "ENOENT") }),
  }).pipe(
    Effect.catchIf(
      ({ isEnoent }) => isEnoent,
      () => Effect.succeed(undefined),
    ),
    Effect.mapError(({ cause }) => new RootEnvReadError({ path: envPath, cause })),
  );

const collectEnvFromProvider = (
  provider: ConfigProvider.ConfigProvider,
  currentPath: readonly string[] = [],
): Effect.Effect<HashMap.HashMap<string, string>, ConfigProvider.SourceError> =>
  Effect.gen(function* () {
    const node = yield* provider.load(currentPath);
    if (!node) {
      return HashMap.empty<string, string>();
    }

    let current = HashMap.empty<string, string>();
    if (node.value !== undefined && currentPath.length > 0) {
      current = HashMap.set(current, currentPath.join("_"), node.value);
    } else if (node._tag === "Value" && currentPath.length > 0) {
      current = HashMap.set(current, currentPath.join("_"), node.value);
    }

    const childSegments: readonly string[] =
      node._tag === "Record"
        ? Array.from(node.keys)
        : node._tag === "Array"
          ? Array.from({ length: node.length }, (_, i) => String(i))
          : [];

    if (childSegments.length === 0) {
      return current;
    }

    const childMaps = yield* Effect.forEach(childSegments, (segment) =>
      collectEnvFromProvider(provider, [...currentPath, segment]),
    );

    return childMaps.reduce((acc, childMap) => HashMap.union(acc, childMap), current);
  });

export const decodeRootEnv = (
  content: string | undefined,
): Effect.Effect<Readonly<Record<string, string>>, ConfigProvider.SourceError> =>
  Effect.gen(function* () {
    if (content === undefined) {
      return Object.freeze({});
    }

    const provider = ConfigProvider.fromDotEnvContents(content, {
      expandVariables: false,
      preserveEmptyStrings: false,
    });

    const hashMap = yield* collectEnvFromProvider(provider);
    return Object.freeze(Object.fromEntries(HashMap.toEntries(hashMap)));
  });

export const loadRootEnv = (
  envPath: string,
): Effect.Effect<Readonly<Record<string, string>>, RootEnvReadError | ConfigProvider.SourceError> =>
  readRootEnvContent(envPath).pipe(Effect.andThen(decodeRootEnv));

export function buildHarness(spec: HarnessSpec): Harness {
  assertPathComponent(spec.sourceName, "harness id");
  const env =
    typeof spec.launcher.env === "function" ? spec.launcher.env(spec.home) : spec.launcher.env;
  return {
    id: spec.id,
    sourceName: spec.sourceName,
    home: spec.home,
    launcher: {
      package: spec.launcher.package,
      bin: spec.launcher.bin,
      distTag: spec.launcher.distTag ?? "latest",
      smokeCheck: spec.launcher.smokeCheck ?? "--version",
      defaultArgs: spec.launcher.defaultArgs ?? [],
      ...(env === undefined ? {} : { env }),
    },
    instructionFile: spec.instructionFile ?? DEFAULT_INSTRUCTION_FILE,
    runtimeSubdir: spec.runtimeSubdir,
    compatManagedEntries: spec.compatManagedEntries ?? [],
    hooks: normalizeHooks(spec.hooks ?? []),
  };
}

export function discoverHarnesses(
  home: string,
  harnessesHome: string,
  platform: HostPlatform = platformFromProcess(),
): readonly Harness[] {
  return HARNESS_ADAPTERS.filter(
    (adapter) =>
      adapter.platforms.includes(platform) && isDirectory(path.join(harnessesHome, adapter.id)),
  ).map((adapter) => {
    for (const segment of adapter.homeSegments) {
      assertPathComponent(segment, `${adapter.id} home segment`);
    }
    return buildHarness({
      ...adapter,
      id: adapter.id,
      sourceName: adapter.id,
      home: path.join(home, ...adapter.homeSegments),
    });
  });
}

export function supportedHarness(
  home: string,
  sourceName: string,
  platform: HostPlatform,
): Harness | undefined {
  const adapter = HARNESS_ADAPTERS.find(
    (candidate) => candidate.id === sourceName && candidate.platforms.includes(platform),
  );
  if (!adapter) {
    return undefined;
  }
  return buildHarness({
    ...adapter,
    id: adapter.id,
    sourceName: adapter.id,
    home: path.join(home, ...adapter.homeSegments),
  });
}

function isDirectory(candidate: string): boolean {
  return existsSync(candidate) && statSync(candidate).isDirectory();
}

function platformFromProcess(): HostPlatform {
  if (process.platform === "darwin" || process.platform === "linux") {
    return process.platform;
  }
  throw new Error(`unsupported platform: ${process.platform}`);
}

function assertPathComponent(value: string, label: string): void {
  if (!PATH_COMPONENT_PATTERN.test(value) || value === "." || value === "..") {
    throw new Error(`invalid ${label}: ${value}`);
  }
}

export const harnessRoot = (harness: Harness): string =>
  harness.runtimeSubdir ? path.join(harness.home, harness.runtimeSubdir) : harness.home;

export function harnessSourceRoot(harness: Harness, harnessesHome: string): string {
  return harness.runtimeSubdir
    ? path.join(harnessesHome, harness.sourceName, harness.runtimeSubdir)
    : path.join(harnessesHome, harness.sourceName);
}

export const harnessInstructionTarget = (harness: Harness): string =>
  path.join(harnessRoot(harness), harness.instructionFile);

export const harnessInstructionFileName = (harness: Harness): string => harness.instructionFile;

export const harnessManagedStatePath = (harness: Harness, managedStateHome: string): string =>
  path.join(managedStateHome, `${harness.sourceName}.json`);

function normalizeHooks(hooks: readonly HarnessHookSpec[]): HarnessHook[] {
  return hooks.map((hook) => {
    switch (hook.kind) {
      case "PackageBootstrap":
        return {
          kind: hook.kind,
          manifestFile: hook.manifestFile ?? "packages.json",
          settingsFile: hook.settingsFile ?? "settings.json",
          cacheSubdir: hook.cacheSubdir ?? DEFAULT_PACKAGE_CACHE_SUBDIR,
        };
      case "ExtensionDeps":
        return {
          kind: hook.kind,
          rootDir: hook.rootDir,
        };
      default:
        return assertNever(hook);
    }
  });
}

File: sync/src/sync/core/harness.py
# Copyright (c) 2026 agents-sync. SPDX-License-Identifier: AGPL-3.0-or-later
"""Harness definitions, environment discovery, and root configuration loader."""

from __future__ import annotations

import io
import os
import re
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import TYPE_CHECKING

import dotenv

from sync.core.harness_adapters import (
    DEFAULT_INSTRUCTION_FILE,
    DEFAULT_PACKAGE_CACHE_SUBDIR,
    HARNESS_ADAPTERS,
    ExtensionDepsHook,
    HarnessAdapter,
    HarnessHookSpec,
    HarnessId,
    HarnessLauncherSpec,
    HostPlatform,
    PackageBootstrapHook,
)

if TYPE_CHECKING:
    from collections.abc import Sequence

SOURCE_AGENT_FILE: str = "HARNESS.md"
INSTALL_TIMEOUT_SECONDS: int = 120
INSTALL_TIMEOUT_MS: int = 120_000
MANAGED_STATE_SUBDIR: str = ".local/share/agents/sync-managed"
SKILLS_DST_DIR: str = "skills"
SKILLS_SOURCE_SUBDIR: str = "current"
PATH_COMPONENT_PATTERN: re.Pattern[str] = re.compile(r"^[A-Za-z0-9._-]+$")

type HarnessHook = PackageBootstrapHook | ExtensionDepsHook


@dataclass(frozen=True, slots=True)
class HarnessLauncher:
    """Resolved harness launcher configuration."""

    package: str
    bin: str
    dist_tag: str = "latest"
    smoke_check: str = "--version"
    default_args: tuple[str, ...] = ()
    env: dict[str, str] | None = None


@dataclass(frozen=True, slots=True)
class HarnessSpec:
    """Input specification for building a resolved Harness."""

    id: HarnessId
    source_name: str
    home: str
    launcher: HarnessLauncherSpec
    instruction_file: str | None = None
    runtime_subdir: str | None = None
    compat_managed_entries: tuple[str, ...] | None = None
    merge_json_files: tuple[str, ...] | None = None
    hooks: tuple[HarnessHookSpec, ...] | None = None


@dataclass(frozen=True, slots=True)
class Harness:
    """Fully resolved harness configuration."""

    id: HarnessId
    source_name: str
    home: str
    launcher: HarnessLauncher
    instruction_file: str = DEFAULT_INSTRUCTION_FILE
    runtime_subdir: str | None = None
    compat_managed_entries: tuple[str, ...] = ()
    merge_json_files: tuple[str, ...] = ()
    hooks: tuple[HarnessHook, ...] = ()


@dataclass(frozen=True, slots=True)
class SyncEnv:
    """Environment layout and configuration for a sync run."""

    home: str
    ssot_home: str
    runtime_home: str
    skills_home: str
    harnesses_home: str
    mcporter_home: str
    summarize_home: str
    managed_state_home: str
    install_timeout_ms: int
    harnesses: tuple[Harness, ...]
    platform: HostPlatform
    root_env: dict[str, str]

    @classmethod
    def from_system(cls) -> SyncEnv:
        """Create a SyncEnv by resolving the home directory from the environment."""
        home = os.environ.get("HOME", "").strip()
        if not home:
            expanded = str(Path("~").expanduser()).strip()
            if expanded and expanded != "~":
                home = expanded
        if not home:
            message = "missing HOME"
            raise RuntimeError(message)
        return cls.from_home(home, INSTALL_TIMEOUT_MS)

    @classmethod
    def from_home(
        cls,
        home: str,
        install_timeout_ms: int = INSTALL_TIMEOUT_MS,
        *,
        platform: HostPlatform | None = None,
    ) -> SyncEnv:
        """Create a SyncEnv for a specified user home directory."""
        home_path = Path(home)
        agents_home = str(home_path / ".config" / "agents")
        harnesses_home = str(home_path / ".config" / "agents" / "harnesses")
        resolved_platform = (
            platform if platform is not None else platform_from_process()
        )
        env_path = str(home_path / ".config" / "agents" / ".env")
        return cls(
            home=home,
            ssot_home=agents_home,
            runtime_home=str(home_path / ".local" / "share" / "agents"),
            skills_home=str(home_path / ".config" / "agents" / "skills"),
            harnesses_home=harnesses_home,
            mcporter_home=str(home_path / ".mcporter"),
            summarize_home=str(home_path / ".summarize"),
            managed_state_home=str(home_path / MANAGED_STATE_SUBDIR),
            install_timeout_ms=install_timeout_ms,
            harnesses=discover_harnesses(home, harnesses_home, resolved_platform),
            platform=resolved_platform,
            root_env=load_root_env(env_path),
        )

    def harness(self, harness_id: HarnessId) -> Harness | None:
        """Look up a discovered harness by its identifier."""
        return next((h for h in self.harnesses if h.id == harness_id), None)


class RootEnvReadError(Exception):
    """Raised when reading the root .env file fails with a non-ENOENT error."""

    path: str
    cause: BaseException | None

    def __init__(self, path: str, cause: BaseException | None = None) -> None:
        """Initialize RootEnvReadError with path and optional cause."""
        self.path = path
        self.cause = cause
        message = f"failed to read root environment file {path}"
        super().__init__(message)


def read_root_env_content(env_path: str) -> str | None:
    """Read the content of the root environment file.

    Returns None if the file does not exist (ENOENT).
    Raises RootEnvReadError if reading fails for any other reason (e.g. EISDIR).
    """
    try:
        return Path(env_path).read_text(encoding="utf-8")
    except FileNotFoundError:
        return None
    except OSError as err:
        raise RootEnvReadError(env_path, err) from err


def decode_root_env(content: str | None) -> dict[str, str]:
    """Decode root .env content into a key-value dictionary.

    - Variable expansion is disabled (interpolate=False).
    - Empty string values are excluded.
    - Purely file-based; values are not merged with or overridden by os.environ.
    """
    if content is None:
        return {}
    raw = dotenv.dotenv_values(stream=io.StringIO(content), interpolate=False)
    return {k: v for k, v in raw.items() if v is not None and v != ""}


def load_root_env(env_path: str) -> dict[str, str]:
    """Load and parse the root .env file at `env_path`.

    Returns an empty dict if the file does not exist.
    Raises RootEnvReadError if reading fails.
    """
    content = read_root_env_content(env_path)
    return decode_root_env(content)


def build_harness(spec: HarnessSpec) -> Harness:
    """Build a resolved Harness instance from a specification."""
    assert_path_component(spec.source_name, "harness id")
    launcher_env: dict[str, str] | None = None
    if callable(spec.launcher.env):
        launcher_env = spec.launcher.env(spec.home)
    elif spec.launcher.env is not None:
        launcher_env = spec.launcher.env

    launcher = HarnessLauncher(
        package=spec.launcher.package,
        bin=spec.launcher.bin,
        dist_tag=(
            spec.launcher.dist_tag if spec.launcher.dist_tag is not None else "latest"
        ),
        smoke_check=(
            spec.launcher.smoke_check
            if spec.launcher.smoke_check is not None
            else "--version"
        ),
        default_args=(
            spec.launcher.default_args if spec.launcher.default_args is not None else ()
        ),
        env=launcher_env,
    )
    return Harness(
        id=spec.id,
        source_name=spec.source_name,
        home=spec.home,
        launcher=launcher,
        instruction_file=(
            spec.instruction_file
            if spec.instruction_file is not None
            else DEFAULT_INSTRUCTION_FILE
        ),
        runtime_subdir=spec.runtime_subdir,
        compat_managed_entries=(
            spec.compat_managed_entries
            if spec.compat_managed_entries is not None
            else ()
        ),
        merge_json_files=(
            spec.merge_json_files if spec.merge_json_files is not None else ()
        ),
        hooks=normalize_hooks(spec.hooks if spec.hooks is not None else ()),
    )


def _adapter_target_home(adapter: HarnessAdapter, user_home: str) -> str:
    for segment in adapter.home_segments:
        assert_path_component(segment, f"{adapter.id} home segment")
    return str(Path(user_home).joinpath(*adapter.home_segments))


def _adapter_to_harness(
    adapter: HarnessAdapter,
    user_home: str,
    source_name: str | None = None,
) -> Harness:
    target_home = _adapter_target_home(adapter, user_home)
    spec = HarnessSpec(
        id=adapter.id,
        source_name=source_name if source_name is not None else adapter.id,
        home=target_home,
        launcher=adapter.launcher,
        instruction_file=adapter.instruction_file,
        runtime_subdir=adapter.runtime_subdir,
        compat_managed_entries=adapter.compat_managed_entries,
        merge_json_files=adapter.merge_json_files,
        hooks=adapter.hooks,
    )
    return build_harness(spec)


def discover_harnesses(
    home: str,
    harnesses_home: str,
    platform: HostPlatform | None = None,
) -> tuple[Harness, ...]:
    """Discover active harnesses based on installed directories and target platform."""
    resolved_platform = platform if platform is not None else platform_from_process()
    harnesses_path = Path(harnesses_home)
    return tuple(
        _adapter_to_harness(adapter, home)
        for adapter in HARNESS_ADAPTERS
        if resolved_platform in adapter.platforms
        and is_directory(str(harnesses_path / adapter.id))
    )


def supported_harness(
    home: str,
    source_name: str,
    platform: HostPlatform,
) -> Harness | None:
    """Find a supported harness adapter by name and platform."""
    for adapter in HARNESS_ADAPTERS:
        if adapter.id == source_name and platform in adapter.platforms:
            return _adapter_to_harness(adapter, home, source_name=source_name)
    return None


def is_directory(candidate: str) -> bool:
    """Check if a filesystem path exists and is a directory."""
    try:
        return Path(candidate).is_dir()
    except OSError:
        return False


def platform_from_process() -> HostPlatform:
    """Determine the host platform from the current Python runtime."""
    system = sys.platform
    if system in ("darwin", "linux"):
        return system
    message = f"unsupported platform: {system}"
    raise RuntimeError(message)


def assert_path_component(value: str, label: str) -> None:
    """Validate that a path segment is safe and matches PATH_COMPONENT_PATTERN."""
    if not PATH_COMPONENT_PATTERN.fullmatch(value) or value in (".", ".."):
        message = f"invalid {label}: {value}"
        raise ValueError(message)


def harness_root(harness: Harness) -> str:
    """Return the absolute root directory of a harness."""
    home_path = Path(harness.home)
    return (
        str(home_path / harness.runtime_subdir)
        if harness.runtime_subdir
        else harness.home
    )


def harness_source_root(harness: Harness, harnesses_home: str) -> str:
    """Return the SSOT source directory of a harness."""
    base_path = Path(harnesses_home) / harness.source_name
    return (
        str(base_path / harness.runtime_subdir)
        if harness.runtime_subdir
        else str(base_path)
    )


def harness_instruction_target(harness: Harness) -> str:
    """Return the destination path for the harness instruction file."""
    return str(Path(harness_root(harness)) / harness.instruction_file)


def harness_instruction_file_name(harness: Harness) -> str:
    """Return the base instruction file name for a harness."""
    return harness.instruction_file


def harness_managed_state_path(harness: Harness, managed_state_home: str) -> str:
    """Return the managed state JSON path for a harness."""
    return str(Path(managed_state_home) / f"{harness.source_name}.json")


def normalize_hooks(
    hooks: Sequence[HarnessHookSpec] = (),
) -> tuple[HarnessHook, ...]:
    """Normalize harness hook specifications with default file paths."""
    result: list[HarnessHook] = []
    for hook in hooks:
        if isinstance(hook, PackageBootstrapHook):
            result.append(
                PackageBootstrapHook(
                    manifest_file=(
                        hook.manifest_file
                        if hook.manifest_file is not None
                        else "packages.json"
                    ),
                    settings_file=(
                        hook.settings_file
                        if hook.settings_file is not None
                        else "settings.json"
                    ),
                    cache_subdir=(
                        hook.cache_subdir
                        if hook.cache_subdir is not None
                        else DEFAULT_PACKAGE_CACHE_SUBDIR
                    ),
                )
            )
        else:
            result.append(hook)
    return tuple(result)


__all__ = [
    "DEFAULT_INSTRUCTION_FILE",
    "DEFAULT_PACKAGE_CACHE_SUBDIR",
    "HARNESS_ADAPTERS",
    "INSTALL_TIMEOUT_MS",
    "INSTALL_TIMEOUT_SECONDS",
    "MANAGED_STATE_SUBDIR",
    "PATH_COMPONENT_PATTERN",
    "SKILLS_DST_DIR",
    "SKILLS_SOURCE_SUBDIR",
    "SOURCE_AGENT_FILE",
    "ExtensionDepsHook",
    "Harness",
    "HarnessAdapter",
    "HarnessHook",
    "HarnessHookSpec",
    "HarnessId",
    "HarnessLauncher",
    "HarnessLauncherSpec",
    "HarnessSpec",
    "HostPlatform",
    "PackageBootstrapHook",
    "RootEnvReadError",
    "SyncEnv",
    "assert_path_component",
    "build_harness",
    "decode_root_env",
    "discover_harnesses",
    "harness_instruction_file_name",
    "harness_instruction_target",
    "harness_managed_state_path",
    "harness_root",
    "harness_source_root",
    "is_directory",
    "load_root_env",
    "platform_from_process",
    "read_root_env_content",
    "supported_harness",
]
